include/meta/meta.hpp

```diff
@@ -925,23 +937,91 @@ namespace meta
-    /// of calling \c Fun with each element in the lists, pairwise.
+    /// in \p List with the unary Metafuncion Class \p Fun. \c transform can
+    /// also be called with two lists of the same length and a binary
+    /// Metafunction Class, in which case it returns a new list constructed with
+    /// the results of calling \c Fun with each element in the lists, pairwise.
     /// \par Complexity
     /// \f$ O(N) \f$.
     template <typename... Args>
     using transform = eval<detail::transform_<Args...>>;
 
+    ////////////////////////////////////////////////////////////////////////////
+    // filter
+    /// \cond
+    namespace detail
+    {
+        template <typename Predicate> struct filter_
+        {
+            template <typename State, typename A>
+            using apply = meta::if_<meta::apply<Predicate, A>,
+                                    meta::push_back<State, A>, State>;
+        };
+    } // namespace detail
+    /// \endcond
+
+    /// Returns a new meta::list where only those elements of \p List A that
+    /// satisfy the Metafunction Class \p Predicate such that
+    /// `apply<Pred,A>::%value` is \c true are present. That is, those elements
+    /// that don't satisfy the \p Predicate are "removed".
+    template <typename List, typename Predicate>
+    using filter = meta::foldl<List, meta::list<>, detail::filter_<Predicate>>;
+
+    ////////////////////////////////////////////////////////////////////////////
+    // for_each
+    /// \cond
+    namespace detail
+    {
+        struct for_each_fn
+        {
+           private:
+            template <class UnaryFunction>
+            constexpr auto impl_(UnaryFunction &&unary_function) const
+            {
+                return std::forward<UnaryFunction>(unary_function);
+            }
+
+            template <class UnaryFunction, class Arg>
+            constexpr auto impl_(UnaryFunction &&unary_function,
+                                 Arg &&arg) const
+            {
+                unary_function(std::forward<Arg>(arg));
+                return impl_(std::forward<UnaryFunction>(unary_function));
+            }
+
+            template <class UnaryFunction, class Arg, class... Args>
+            constexpr auto impl_(UnaryFunction &&unary_function, Arg &&arg,
+                                 Args &&... args) const
+            {
+                unary_function(std::forward<Arg>(arg));
+                return impl_(std::forward<UnaryFunction>(unary_function),
+                             std::forward<Args>(args)...);
+            }
+
+           public:
+            template <class UnaryFunction, class... Args>
+            constexpr auto operator()(meta::list<Args...>,
+                                      UnaryFunction &&unary_function) const
+            {
+                return impl_(std::forward<UnaryFunction>(unary_function),
+                             Args{}...);
+            }
+        };
+    } // namespace detail
+    /// \endcond
```

Comment:
This is not valid C++11. I'd like to keep Meta usable in C++11 mode. Also, is there some reason you did it recursively instead of iteratively, like https://twitter.com/ericniebler/status/559119062895431680